File: Java/TcpServer.java
import java.io.*;
import java.net.*;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

// =========================================================
// CLIENT HANDLER
// =========================================================

class ClientHandler {

    String name;
    Socket socket;
    BufferedReader in;
    PrintWriter out;

    ClientHandler(
            String name,
            Socket socket,
            BufferedReader in,
            PrintWriter out
    ) {
        this.name = name;
        this.socket = socket;
        this.in = in;
        this.out = out;
    }
}

// =========================================================
// TCP SERVER
// =========================================================

public class TcpServer {

    private static final Map<String, ClientHandler> clients =
            new ConcurrentHashMap<>();

    public static void main(String[] args) {

        start(5000);
    }

    public static void start(int port) {

        try (ServerSocket serverSocket =
                     new ServerSocket(port)) {

            System.out.println(
                    "TCP Server läuft auf Port " + port
            );

            while (true) {

                Socket socket = serverSocket.accept();

                new Thread(() ->
                        handle(socket)
                ).start();
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // =====================================================
    // CLIENT THREAD
    // =====================================================

    private static void handle(Socket socket) {

        String name = null;

        try (

                BufferedReader in = new BufferedReader(
                        new InputStreamReader(
                                socket.getInputStream()
                        )
                );

                PrintWriter out = new PrintWriter(
                        socket.getOutputStream(),
                        true
                )

        ) {

            // =================================================
            // REGISTRIERUNG
            // =================================================

            while (true) {

                out.println("register <name>");

                String reg = in.readLine();

                if (reg == null) {
                    socket.close();
                    return;
                }

                String[] parts = reg.split(" ");

                // =============================================
                // FORMAT PRÜFEN
                // =============================================

                if (parts.length != 2 ||
                        !parts[0].equalsIgnoreCase("register")) {

                    out.println(
                            "Erwartet: register <name>"
                    );

                    continue;
                }

                name = parts[1];

                // =============================================
                // NAME SCHON VERGEBEN
                // =============================================

                if (clients.containsKey(name)) {

                    out.println(
                            "Name bereits vergeben!"
                    );

                    continue;
                }

                break;
            }

            // =================================================
            // CLIENT REGISTRIEREN
            // =================================================

            ClientHandler handler =
                    new ClientHandler(
                            name,
                            socket,
                            in,
                            out
                    );

            clients.put(name, handler);

            System.out.println(name + " connected");

            // =================================================
            // MESSAGE LOOP
            // =================================================

            String msg;

            while ((msg = in.readLine()) != null) {

                System.out.println(name + ": " + msg);

                // =============================================
                // SEND
                // =============================================

                if (msg.startsWith("send ")) {

                    String[] parts =
                            msg.split(" ", 3);

                    if (parts.length == 3) {

                        String target = parts[1];
                        String text = parts[2];

                        if (clients.containsKey(target)) {

                            clients.get(target).out.println(
                                    name + ": " + text
                            );
                        }
                    }
                }
            }

        } catch (Exception e) {

            System.out.println(name + " disconnected");

        } finally {

            // =================================================
            // DISCONNECT
            // =================================================

            if (name != null) {

                clients.remove(name);

                System.out.println(
                        "Removed: " + name
                );
            }

            try {
                socket.close();
            } catch (IOException ignored) {}
        }
    }
}

File: Python/TcpServer.py
import socket
import threading

# =========================================================
# CLIENT HANDLER
# =========================================================

class ClientHandler:

    def __init__(self, name, conn):
        self.name = name
        self.conn = conn


# =========================================================
# CLIENT MAP
# =========================================================

clients = {}


# =========================================================
# CLIENT THREAD
# =========================================================

def handle_client(conn):

    name = None

    try:

        # =====================================================
        # REGISTRIERUNG
        # =====================================================

        while True:

            conn.sendall(b"register <name>\n")

            reg = conn.recv(1024).decode().strip()

            if not reg:
                conn.close()
                return

            parts = reg.split(" ")

            # ===== FORMAT PRUEFEN =====

            if len(parts) != 2 or parts[0].lower() != "register":

                conn.sendall(
                    b"Erwartet: register <name>\n"
                )

                continue

            name = parts[1]

            # ===== NAME SCHON VERGEBEN =====

            if name in clients:

                conn.sendall(
                    b"Name bereits vergeben!\n"
                )

                continue

            break

        # =====================================================
        # CLIENT REGISTRIEREN
        # =====================================================

        clients[name] = ClientHandler(name, conn)

        print(f"{name} connected")

        # =====================================================
        # MESSAGE LOOP
        # =====================================================

        while True:

            data = conn.recv(1024)

            if not data:
                break

            msg = data.decode().strip()

            print(f"{name}: {msg}")

            # =================================================
            # SEND
            # =================================================

            if msg.startswith("send "):

                parts = msg.split(" ", 2)

                if len(parts) == 3:

                    target = parts[1]
                    text = parts[2]

                    if target in clients:

                        clients[target].conn.sendall(
                            f"{name}: {text}\n".encode()
                        )

                continue

    except Exception as e:

        print(f"{name} disconnected")

    finally:

        # =====================================================
        # DISCONNECT
        # =====================================================

        if name is not None:

            clients.pop(name, None)

            print(f"Removed: {name}")

        conn.close()


# =========================================================
# TCP SERVER
# =========================================================

def tcp_server(port=5000):

    server = socket.socket(
        socket.AF_INET,
        socket.SOCK_STREAM
    )

    server.bind(("0.0.0.0", port))

    server.listen()

    print(f"TCP Server läuft auf Port {port}")

    while True:

        conn, addr = server.accept()

        threading.Thread(
            target=handle_client,
            args=(conn,)
        ).start()


# =========================================================
# MAIN
# =========================================================

if __name__ == "__main__":

    tcp_server()